[DDW-741] Implement Wallet sorting

renanvalentin opened the pull request:
This PR introduces wallet sorting on the sidebar menu.

## Todos

- [ ] Todo

## Screenshots

![Screen Shot 2021-12-08 at 09 39 59](https://user-images.githubusercontent.com/2846918/145209859-207a426a-5a6e-43d0-bec4-f01ebaa8e9ef.png)
![Screen Shot 2021-12-08 at 09 36 36](https://user-images.githubusercontent.com/2846918/145209910-00cae980-10ec-4a78-9241-704e1a490ba9.png)
![Screen Shot 2021-12-08 at 09 36 48](https://user-images.githubusercontent.com/2846918/145209916-5e57845f-5d65-4332-bb2f-98baed1e5699.png)
![Screen Shot 2021-12-08 at 09 37 18](https://user-images.githubusercontent.com/2846918/145209918-3994cefa-5725-4ecf-a5b0-527d696d76b1.png)
![Screen Shot 2021-12-09 at 16 38 58](https://user-images.githubusercontent.com/2846918/145464338-aea5431b-60eb-4ff9-ad19-9633e5d3a398.png)



---

## Testing Checklist

- [Slack QA thread](https://input-output-rnd.slack.com/archives/GGKFXSKC6/p1638968067385300)
- [ ] Test

---

## Test scenarios 

**Scenario 1 - Validate sorting buttons**
```gherkin
Given that Daedalus has at least one wallet
When I check list of the wallets 
Then I can see that the following sorting buttons are displayed at the bottom:

- Date
- Balance
- A - Z
```
**Scenario 2 - Validate default sorting**
```gherkin
Given that Daedalus has at least two wallets
And I didn't change the sorting of the list
When I check list of the wallets 
Then I can see that the default sorting is creation date, showing the newly added at the bottom of the list
And I can see that Shelley and Byron wallets are sorted independently, showing Byron wallets at the bottom
And at the bottom of the list I can see the date button is selected
And it has an arrow the indicates the sorting is descending 
```
**Scenario 3 - Validate explanation tooltips**
```gherkin
Given that Daedalus has at least one wallets
When I check list of the wallets 
And I hover on the sorting buttons 
Then I can see that an explanation tooltip is displayed for each button
And it indicates which criteria applies for the corresponding option:
Sort wallets by [criteria]
```
**Scenario 4 - Sort wallets by date**
```gherkin
Given that Daedalus has at least two wallets
And the list is not sorted by date
When I click on "Date" at the bottom of the list
Then I can see that the wallets are sorted by the date of creation/restoration,  showing the first added at the bottom of the list
And upon clicking the "date" button again, the order of the list is inverted
And the arrow on the button is inverted
```
**Scenario 5 - Sort wallets by name**
```gherkin
Given that Daedalus has at least two wallets
And the list is not sorted by name
When I click on "A-Z" at the bottom of the list
Then I can see that the wallets are sorted alphabetically
And upon clicking the "A-Z" button again, the order of the list is inverted
And the arrow on the button is inverted
```
**Scenario 6 - Sort wallets by balance**
```gherkin
Given that Daedalus has at least two wallets
And the list is not sorted by balance
When I click on "Balance" at the bottom of the list
Then I can see that the wallets are sorted by their balance, showing those with higher balance first
And upon clicking the "Balance" button again, the order of the list is inverted
And the arrow on the button is inverted
```
**Scenario 7 - Sort wallet with same name**
```gherkin
Given that Daedalus has at least two wallets
And the wallets have the same name, but different balance
And the list is not sorted by name
When I click on "A-Z" at the bottom of the list
Then I can see within the wallets with the same name, the one with higher balance are shown first
And upon clicking the "A-Z" button again, the order of the list is inverted
And the arrow on the button is inverted
```
**Scenario 8 - Sort wallet with same balance**
```gherkin
Given that Daedalus has at least two wallets
And the wallets have the same balance, but different name
And the list is not sorted by balance
When I click on "Balance" at the bottom of the list
Then I can see the wallets with the same balance are sorted alphabetically
And upon clicking the "A-Z" button again, the order of the list is inverted
And the arrow on the button is inverted
```
**Scenario 9 - Sort wallet with same balance and name**
```gherkin
Given that Daedalus has at least two wallets
And the wallets have the same balance and name
And the list is not sorted by balance or name
When I click on "Balance" or "A-Z" at the bottom of the list
Then I can see that the wallets are sorted by the date of creation/restoration,  showing the first added at the bottom of the list
And upon clicking the sorting button again, the order of the list is inverted
And the arrow on the button is inverted
```

**Scenario 10 - Search wallets**
```gherkin
Given that Daedalus has at least two wallets
When I type on the “filter” a value that partially matches the name of one of the wallets
Then I can see that only the wallets that matches the filter are listed
And upon clicking on X, the filter is cleared and all the wallets are listed
```

---

## Review Checklist

### Basics

- [x] PR has been assigned and has appropriate labels (`feature`/`bug`/`chore`, `release-x.x.x`)
- [ ] PR is updated to the most recent version of the target branch (and there are no conflicts)
- [ ] PR has a good description that summarizes all changes
- [ ] PR has default-sized Daedalus window screenshots or animated GIFs of important UI changes:
  - [ ] In English
  - [ ] In Japanese
- [ ] CHANGELOG entry has been added to the top of the appropriate section (*Features*, *Fixes*, *Chores*) and is linked to the correct PR on GitHub
- [ ] Automated tests: All acceptance and unit tests are passing (`yarn test`)
- [ ] Manual tests (minimum tests should cover newly added feature/fix): App works correctly in *development* build (`yarn dev`)
- [ ] Manual tests (minimum tests should cover newly added feature/fix): App works correctly in *production* build (`yarn package` / CI builds)
- [ ] There are no *flow* errors or warnings (`yarn flow:test`)
- [ ] There are no *lint* errors or warnings (`yarn lint`)
- [ ] There are no *prettier* errors or warnings (`yarn prettier:check`)
- [ ] There are no missing translations (running `yarn manage:translations` produces no changes)
- [ ] Text changes are proofread and approved (Jane Wild / Amy Reeve)
- [ ] Japanese text changes are proofread and approved (Junko Oda)
- [ ] UI changes look good in all themes (Alexander Rukin)
- [ ] Storybook works and no stories are broken (`yarn storybook`)
- [ ] In case of dependency changes `yarn.lock` file is updated

### Code Quality
- [ ] Important parts of the code are properly commented and documented
- [ ] Code is properly typed with flow
- [ ] React components are split-up enough to avoid unnecessary re-renderings
- [ ] Any code that only works in main process is neatly separated from components

### Testing
- [ ] New feature/change is covered by acceptance tests
- [ ] New feature/change is manually tested and approved by QA team
- [ ] All existing acceptance tests are still up-to-date
- [ ] New feature/change is covered by Daedalus Testing scenario
- [ ] All existing Daedalus Testing scenarios are still up-to-date

### After Review
- [ ] Merge the PR
- [ ] Delete the source branch
- [ ] Move the ticket to `done` column on the YouTrack board
- [ ] Update Slack QA thread by marking it with a green checkmark

przemyslaw-wlodek commented on source/renderer/app/components/sidebar/wallets/WalletSearch.js:
```diff
@@ -0,0 +1,61 @@
+import { noop } from 'lodash/fp';
+import classNames from 'classnames';
+import { Input } from 'react-polymorph/lib/components/Input';
+import SVGInline from 'react-svg-inline';
```
This package is deprecated and they suggest using [SVGR](https://github.com/gregberge/svgr).

In my opinion we could even create a simple wrapper around svg by hand and it works without any external dependencies.

I don't want to bring any confusion with this suggestion but let's discuss what is worth.
I see that `react-svg-inline` is not heavily used in our codebase so we might get rid of it easily (as it's deprecated).

renanvalentin replied:
Good catch! I think this would fit better in a different PR as the current code base heavily relies on `react-svg-inline`. Would you like to create a tech debt ticket for this one?

przemyslaw-wlodek replied:
Sure :)

https://input-output.atlassian.net/browse/DDW-919

przemyslaw-wlodek commented on source/renderer/app/components/sidebar/wallets/WalletSearch.scss:
```diff
@@ -0,0 +1,71 @@
+    border-bottom-color: rgba(250, 251, 252, 0.5);
+
+    .searchIcon {
+      path {
+        fill: var(--theme-sidebar-menu-add-button-text-color);
+      }
```
Ideally we could modify icon colors with just `color` css property e.g. if we change `fill="currentColor"` in svgs.

przemyslaw-wlodek commented on source/renderer/app/stores/SidebarStore.js:
```diff
@@ -62,6 +76,57 @@ export default class SidebarStore extends Store {
+      (w: Wallet, index: number) => ({ wallet: w, index })
+    );
+
+    const amountWeight = ({ wallet }: IndexedWallet) =>
```
1. micro-optimization but all those getters / pickers are recreated with each `_sortWallets` call.
2. can we name them like a functions instead of values e.g. `getAmountWeight` or `pickAmountWeight` ?

przemyslaw-wlodek commented on source/renderer/app/types/sidebarTypes.js:
```diff
@@ -12,3 +12,24 @@ export type SidebarWalletType = {
+
+export type WalletSortOrderOptions = 'asc' | 'desc';
+
+export const WalletSortBy: EnumMap<string, WalletSortByOptions> = Object.freeze(
```
I think we might not need `Object.freeze` as long as we have typing system (flow) and common sense (trust in developers).

`Object.freeze` will work in this case for now but it only gives us a shallow immutability.

In TypeScript there are additional 3rd party helpers that allow `DeepReadonly` which makes it safe without extra code in the bundle.

Let me know your thoughts.

renanvalentin replied:
I guess we can remove it.

alexander-rukin commented:
<img width="188" alt="image" src="https://user-images.githubusercontent.com/2539345/145385242-9745d561-76ed-4991-a402-efc10e70c279.png">

inactive element is not correct - whole "button" should have 0.3 opacity 
and default secondary button hover effects are applied to all these "buttons" - https://zpl.io/2pOlAqE
by "default" I mean: 
- 0.1 - normal 
- 0.05 - hover 
- 0.12 - pressed

alexander-rukin commented on source/renderer/app/i18n/locales/en-US.json:
```diff
@@ -357,6 +357,13 @@
+  "sidebar.wallets.sortByBalanceTooltip": "Sort by balance",
+  "sidebar.wallets.sortByDateButton": "Date",
+  "sidebar.wallets.sortByDateTooltip": "Sort by creation date",
+  "sidebar.wallets.sortByNameButton": "A-Z",
```
what type of dash is here? "-" / "–" / "—" ? 

<img width="45" alt="image" src="https://user-images.githubusercontent.com/2539345/145386068-1a0f5977-11bf-4c43-8996-9147d8e10ac1.png">

it feels too small

renanvalentin replied:
Changed to `A – Z`

alexander-rukin commented on source/renderer/app/components/sidebar/wallets/WalletSearch.scss:
```diff
@@ -0,0 +1,71 @@
+.component {
+  align-items: center;
+
+  border-bottom: 1px solid rgba(250, 251, 252, 0.1);
```
this is light-blue specific color -> should use some variable I suppose

alexander-rukin commented on source/renderer/app/components/sidebar/wallets/WalletSearch.scss:
```diff
@@ -0,0 +1,71 @@
+  &:focus,
+  &:active,
+  &:focus-within {
+    border-bottom-color: rgba(250, 251, 252, 0.5);
```
this is light-blue specific color -> should use some variable I suppose

alexander-rukin commented on source/renderer/app/components/sidebar/wallets/WalletSearch.scss:
```diff
@@ -0,0 +1,71 @@
+    padding-left: 25px;
+
+    &:focus {
+      border-bottom-color: rgba(250, 251, 252, 0.5);
```
this is light-blue specific color -> should use some variable I suppose

alexander-rukin commented on source/renderer/app/themes/daedalus/cardano.js:
```diff
@@ -798,6 +798,9 @@ export const CARDANO_THEME_OUTPUT = {
     '--theme-sidebar-category-networkInfo-text-color': '#fff',
+    '--theme-sidebar-sort-button-active-background-color':
+      'rgba(250, 251, 252, 0.1)',
+    '--theme-sidebar-sort-button-background-color': 'rgba(250, 251, 252, 0.03)',
```
this is redundant, as the whole object just has 0.3 opacity in this case

alexander-rukin commented on source/renderer/app/themes/daedalus/dark-blue.js:
```diff
@@ -801,6 +801,9 @@ export const DARK_BLUE_THEME_OUTPUT = {
     '--theme-sidebar-category-networkInfo-background-color': '#536370',
     '--theme-sidebar-category-networkInfo-text-color': '#e9f4fe',
+    '--theme-sidebar-sort-button-active-background-color':
+      'rgba(250, 251, 252, 0.1)',
```
this color is incorrect (it's light-blue theme color) 
it should be rgba (233, 244, 254, 0.1)

alexander-rukin commented on source/renderer/app/themes/daedalus/dark-cardano.js:
```diff
@@ -785,6 +785,9 @@ export const DARK_CARDANO_THEME_OUTPUT = {
     '--theme-sidebar-category-networkInfo-background-color': '#1fc1c3',
     '--theme-sidebar-category-networkInfo-text-color': '#121326',
+    '--theme-sidebar-sort-button-active-background-color':
+      'rgba(250, 251, 252, 0.1)',
```
this color is incorrect (it's light-blue theme color) 
it should be rgba (255, 255, 255, 0.1)

alexander-rukin commented on source/renderer/app/themes/daedalus/flight-candidate.js:
```diff
@@ -783,6 +783,9 @@ export const FLIGHT_CANDIDATE_THEME_OUTPUT = {
     '--theme-sidebar-category-networkInfo-background-color': '#ffb923',
     '--theme-sidebar-category-networkInfo-text-color': '#121326',
+    '--theme-sidebar-sort-button-active-background-color':
+      'rgba(250, 251, 252, 0.1)',
```
this color is incorrect (it's light-blue theme color) 
it should be rgba (255, 255, 255, 0.1)

alexander-rukin commented on source/renderer/app/themes/daedalus/incentivized-testnet.js:
```diff
@@ -785,6 +785,9 @@ export const INCENTIVIZED_TESTNET_THEME_OUTPUT = {
     '--theme-sidebar-category-networkInfo-background-color': '#eb2256',
     '--theme-sidebar-category-networkInfo-text-color': '#121326',
+    '--theme-sidebar-sort-button-active-background-color':
+      'rgba(250, 251, 252, 0.1)',
```
this color is incorrect (it's light-blue theme color) 
it should be rgba (255, 255, 255, 0.1)

alexander-rukin commented on source/renderer/app/themes/daedalus/shelley-testnet.js:
```diff
@@ -784,6 +784,9 @@ export const SHELLEY_TESTNET_THEME_OUTPUT = {
     '--theme-sidebar-category-networkInfo-background-color': '#898ee6',
     '--theme-sidebar-category-networkInfo-text-color': '#121326',
+    '--theme-sidebar-sort-button-active-background-color':
+      'rgba(250, 251, 252, 0.1)',
```
this color is incorrect (it's light-blue theme color) 
it should be rgba (255, 255, 255, 0.1)

alexander-rukin commented on source/renderer/app/themes/daedalus/white.js:
```diff
@@ -788,6 +788,9 @@ export const WHITE_THEME_OUTPUT = {
     '--theme-sidebar-category-networkInfo-background-color': '#29b595',
     '--theme-sidebar-category-networkInfo-text-color': '#fff',
+    '--theme-sidebar-sort-button-active-background-color':
+      'rgba(250, 251, 252, 0.1)',
```
this color is incorrect (it's light-blue theme color) 
it should be rgba (45, 45, 45, 0.1)

alexander-rukin commented on source/renderer/app/themes/daedalus/yellow.js:
```diff
@@ -785,6 +785,9 @@ export const YELLOW_THEME_OUTPUT = {
     '--theme-sidebar-category-networkInfo-background-color': '#2d2d2d',
     '--theme-sidebar-category-networkInfo-text-color': '#fff',
+    '--theme-sidebar-sort-button-active-background-color':
+      'rgba(250, 251, 252, 0.1)',
```
this color is incorrect (it's light-blue theme color) 
it should be rgba (45, 45, 45, 0.1)

alexander-rukin commented:
<img width="297" alt="image" src="https://user-images.githubusercontent.com/2539345/145387573-dd2b1618-4946-4764-9248-0f72e107273a.png">

- height of search area equal to wallet line - https://zpl.io/a7jmPPK 
- colors are not right 

<img width="229" alt="image" src="https://user-images.githubusercontent.com/2539345/145387701-2199f80d-e579-4223-b3b5-a71c134311b3.png">

- punctuation check should be off I suppose 
- "x" icon use wrong colors 

<img width="206" alt="image" src="https://user-images.githubusercontent.com/2539345/145387848-64b19c04-6db2-434d-91ce-74d186877256.png">

- scroll issues 🤔

przemyslaw-wlodek commented on source/renderer/app/components/sidebar/wallets/SidebarWalletsMenu.js:
```diff
@@ -40,6 +85,47 @@ export default class SidebarWalletsMenu extends Component<Props> {
+    });
+  };
+
+  getWallets = (
```
or even better `filterWalletsBySearchValue`

renanvalentin commented on source/renderer/app/components/sidebar/wallets/SidebarWalletsMenu.js:
```diff
@@ -40,6 +85,47 @@ export default class SidebarWalletsMenu extends Component<Props> {
+    wallets: SidebarWalletType[]
+  ): SidebarWalletType[] => {
+    if (searchValue.length > 0) {
+      return wallets.filter((w) =>
+        w.title.toLowerCase().includes(searchValue.toLowerCase())
+      );
```
This would be a good candidate for analytics. Maybe we could start with a simpler version and see the community feedback around this new feature.

DominikGuzei commented on source/renderer/app/stores/SidebarStore.js:
```diff
@@ -19,6 +31,8 @@ export default class SidebarStore extends Store {
     );
     sidebarActions.walletSelected.listen(this._onWalletSelected);
+    sidebarActions.changeWalletSortType.listen(this._onChangeWalletSortType);
+    sidebarActions.searchValueUpdated.listen(this._onSearchValueUpdated);
```
We discussed in one of our dev calls that we figured that we want to move away from using the "Action" pattern because it doesn't add any benefits over public class methods. So in this case you could remove both actions and also remove the `_` in front of the class methods + call them directly from within the components.

DominikGuzei commented on source/renderer/app/stores/SidebarStore.spec.js:
```diff
@@ -0,0 +1,180 @@
+
+import SidebarStore from './SidebarStore';
+
+describe('discreetWalletAmount replacer', () => {
```
Copy mistake in the test description 😉

DominikGuzei commented on source/renderer/app/stores/SidebarStore.js:
```diff
@@ -62,6 +80,50 @@ export default class SidebarStore extends Store {
     });
   }
 
+  _sortWallets = (wallets: Wallet[]): Wallet[] => {
```
This could be moved out of the store into a unit-testable helper function. That would reduce the store code also simplify the specs since you wouldn't have to set up the store for each case

DominikGuzei commented on source/renderer/app/stores/SidebarStore.spec.js:
```diff
@@ -0,0 +1,180 @@
+    sidebarStore.walletSortConfig.sortBy = sortBy;
+    sidebarStore.walletSortConfig.sortOrder = sortOrder;
+
+    return sidebarStore._sortWallets((wallets: any));
```
If we really want store integration tests then this would not be enough because we would have to use the public store API, not private methods that might never be called correctly using the public api

DominikGuzei reviewed (changes requested):
Great work, I added a few suggestions on how we can already start improving the existing code base by moving away from "Actions" to public store methods + simplifying the store logic by introducing unit-testable helper functions to do 80% of the algorithmic work and keep stores just as state coordinators

miorsufianiohk commented:
Hi @renanvalentin ,

On build 20071, I found out that:

When I created 2 new empty shelley wallets with the same name, upon sorting by name, they always remained at the top or at the bottom. [Please see video](https://drive.google.com/file/d/12BB7bM5YAWEI0L2YU_ef71w5e7WyC6fj/view?usp=sharing)

When I changed one of their names to something else, they remained in the same position. [Please see video](https://drive.google.com/file/d/1QpAlxaaG4kP1a1G_6xRlpTDRuURHVO7t/view?usp=sharing)

When I changed one of the wallets first letter to another letter, the position remained the same. [Please see video](https://drive.google.com/file/d/1XccHmns874SwVSC7DPUmxFUXFoReYntg/view?usp=sharing)

This is not intentional, isn't it? 🤔

renanvalentin commented:
> Hi @renanvalentin ,
> 
> On build 20071, I found out that:
> 
> When I created 2 new empty shelley wallets with the same name, upon sorting by name, they always remained at the top or at the bottom. [Please see video](https://drive.google.com/file/d/12BB7bM5YAWEI0L2YU_ef71w5e7WyC6fj/view?usp=sharing)
> 
> When I changed one of their names to something else, they remained in the same position. [Please see video](https://drive.google.com/file/d/1QpAlxaaG4kP1a1G_6xRlpTDRuURHVO7t/view?usp=sharing)
> 
> When I changed one of the wallets first letter to another letter, the position remained the same. [Please see video](https://drive.google.com/file/d/1XccHmns874SwVSC7DPUmxFUXFoReYntg/view?usp=sharing)
> 
> This is not intentional, isn't it? 🤔

@miorsufianiohk good catch! I added another test case to cover your scenario:
https://github.com/input-output-hk/daedalus/pull/2775/commits/54b72e18973418d13ea04a1cd39865833b8a38e5#diff-2702d42f5ab9e42dd45eb9e82aaad4b81f20d820c00fed4f48a59d0d88ab64ceR94

gabriela-ponce commented:
@renanvalentin Here is an issue I found on the filter:

- Insert a value that matches one of the wallets, then insert some additional characters that are not in the wallet name. You'll see the wallet is still listed, until you insert certain amount of characters that do not match. Check [this](https://drive.google.com/file/d/1mw-Q9Y2D00IfbGTHwk62xqd62D-SnDMb/view?usp=sharing) video.

miorsufianiohk commented:
Hi @renanvalentin,

On build 20085:

Thanks for fixing [this issue](https://github.com/input-output-hk/daedalus/pull/2775#issuecomment-993425774). Looks good now. 

In addition to what @gabriela-ponce found [here](https://github.com/input-output-hk/daedalus/pull/2775#issuecomment-994054173), I'd also like to add that when I pressed space bar, one of the wallets appeared in the search. It didn't seem right to me. Please [see video](https://drive.google.com/file/d/1ndC5BX-wK7kvt8rFENIalz_G0r1S8fZ-/view?usp=sharing).

miorsufianiohk commented:
Hi @renanvalentin ,

On build 20214, I still could reproduce the filter issues that @gabriela-ponce and I found

- Upon entering a space character, only some wallets were displayed eventhough there were a few other wallets with spaces in their names. [See video 1](https://drive.google.com/file/d/1P8PMyRQocMwlMuVOmtzEaa5WOTnK6IJG/view?usp=sharing)
- After entering the first few characters from a wallet name and then entering additional spaces, I observed the result was still showing these wallets. However, after a few more additional spaces were entered, the wallets disappeared. [See video 2](https://drive.google.com/file/d/1Lw9Ci61LnLnRvcrPRsq7KDO2hwUyI6li/view?usp=sharing)
- After entering characters that existed in the wallet name, when entering a non-existing character, the wallet was still displayed. [See video 3](https://drive.google.com/file/d/1SDnACSpz1h9hteD0Eiv8IY53JINWm4Qk/view?usp=sharing)

renanvalentin commented:
> Hi @renanvalentin ,
> 
> On build 20214, I still could reproduce the filter issues that @gabriela-ponce and I found
> 
> * Upon entering a space character, only some wallets were displayed eventhough there were a few other wallets with spaces in their names. [See video 1](https://drive.google.com/file/d/1P8PMyRQocMwlMuVOmtzEaa5WOTnK6IJG/view?usp=sharing)
> * After entering the first few characters from a wallet name and then entering additional spaces, I observed the result was still showing these wallets. However, after a few more additional spaces were entered, the wallets disappeared. [See video 2](https://drive.google.com/file/d/1Lw9Ci61LnLnRvcrPRsq7KDO2hwUyI6li/view?usp=sharing)
> * After entering the first few characters and then a non-existent character, I observed the wallet was still showing. Also, when I continued entering this non-existent character, the wallet remain displayed. [See video 3](https://drive.google.com/file/d/1ynbRXT_b20-G5JgJVaFlliQsSYKjfUkv/view?usp=sharing)

Hi @miorsufianiohk that's expected unless we want to switch from fuzzy-search (which gives a score to each items based on some parameters) to a more strict one where each character must match 1:1. 

> * Upon entering a space character, only some wallets were displayed eventhough there were a few other wallets with spaces in their names. [See video 1](https://drive.google.com/file/d/1P8PMyRQocMwlMuVOmtzEaa5WOTnK6IJG/view?usp=sharing)

In this case `empty space` matched `My Trezor` & `byron new` because both have wording space. You will notice no result has been found once 2 empty spaces were added. This is due to wallet names not having this pattern.

> * After entering the first few characters from a wallet name and then entering additional spaces, I observed the result was still showing these wallets. However, after a few more additional spaces were entered, the wallets disappeared. [See video 2](https://drive.google.com/file/d/1Lw9Ci61LnLnRvcrPRsq7KDO2hwUyI6li/view?usp=sharing)

This is like the example above where `byron      ` matched the keywords `byron` & `byron new`

miorsufianiohk commented:
Hi @renanvalentin,

This is the findings on build 20225

Thank you for fixing the issue when entering space as the first few characters, looks good now 👍 .

However, these issues still remain:

- I had wallets named "byron" and "byron new new1". When I entered "byron" and a "1", both wallets still displayed. When I entered another "1", only "byron" wallet was displayed eventhough byron didn't have a "1" in it. Please [see video 1](https://drive.google.com/file/d/1s6zml_oFlpAUwpFs80BQBC5OWT32yLjG/view?usp=sharing)
- I had wallets named "byron" and "byron new new1". When I entered "byron" and a few spaces, the wallet "byron" remained being shown eventhough wallet "byron" didn't even have a space in its name. Please [see video 2](https://drive.google.com/file/d/1Vg8WG039aqQMBViMhs_27Sg-Y4_9xbBZ/view?usp=sharing)

renanvalentin commented:
> Hi @renanvalentin,
> 
> This is the findings on build 20225
> 
> Thank you for fixing the issue when entering space as the first few characters, looks good now 👍 .
> 
> However, these issues still remain:
> 
> * I had wallets named "byron" and "byron new new1". When I entered "byron" and a "1", both wallets still displayed. When I entered another "1", only "byron" wallet was displayed eventhough byron didn't have a "1" in its name. Please [see video 1](https://drive.google.com/file/d/1s6zml_oFlpAUwpFs80BQBC5OWT32yLjG/view?usp=sharing)
> * I had wallets named "byron" and "byron new new1". When I entered "byron" and a few spaces, the wallet "byron" remained being shown eventhough wallet "byron" didn't even have a space in its name. Please [see video 2](https://drive.google.com/file/d/1Vg8WG039aqQMBViMhs_27Sg-Y4_9xbBZ/view?usp=sharing)

@miorsufianiohk fix should be good by now 🚀

alexander-rukin commented:
<img width="181" alt="image" src="https://user-images.githubusercontent.com/2539345/149168351-626df570-55c9-4480-93a9-29adce90990c.png">
for this "x" icon -> normal state is 0.5 opacity -> hover state is 1 opacity

alexander-rukin commented:
<img width="195" alt="image" src="https://user-images.githubusercontent.com/2539345/149169009-81bda7b8-a833-44ca-a77d-5bb165a54062.png">
corners of these badges should be sensitive to hover how hover area seems smaller

alexander-rukin commented:
<img width="78" alt="image" src="https://user-images.githubusercontent.com/2539345/149169297-18099222-81ad-4642-bcb7-fe6f88b5ad88.png">
text feels not vertically centered

alexander-rukin commented:
<img width="968" alt="image" src="https://user-images.githubusercontent.com/2539345/149169699-7fa96394-18e0-445d-9042-67e317851eb7.png">

arrow icon seems to be bigger and more shifted to the right 
p.s. left - live, right - design

renanvalentin commented:
@alexander-rukin should be looking good now:
https://user-images.githubusercontent.com/2846918/149210892-b35a4260-790f-444a-8f6d-9493e67f9656.mp4

miorsufianiohk reviewed (approved):
LGTM. Tested on 20302. Great work @renanvalentin 👍

alexander-rukin commented:
<img width="1096" alt="image" src="https://user-images.githubusercontent.com/2539345/149385207-d467d9db-f03e-43d1-a98c-a264266b842a.png">
filter input field is not vertically centered

alexander-rukin reviewed (approved):
LGTM!

gabriela-ponce reviewed (approved):
Good job 🎉

lucas-barros commented on source/renderer/app/components/sidebar/wallets/WalletSearch.js:
```diff
@@ -0,0 +1,64 @@
+    >
+      <SVGInline
+        svg={searchIcon}
+        className={classNames({
```
No need for classNames here

renanvalentin commented on source/renderer/app/components/sidebar/types.js:
```diff
@@ -1,13 +1,25 @@
   wallets: ?{
     items: Array<SidebarWalletType>,
     activeWalletId: ?string,
     actions: {
```
@mchappell good catch! The optional chain on `menus?.wallets?.actions.onWalletSortBy` is not needed!

renanvalentin commented on source/renderer/app/components/sidebar/wallets/SidebarWalletsMenu.js:
```diff
@@ -40,6 +87,47 @@ export default class SidebarWalletsMenu extends Component<Props> {
+        }
+      );
+
+      const result = fuse.search(searchValue);
+
+      const filteredResult = result.map((r) => r.item);
+
+      return filteredResult;
```
Yeah, makes sense! I've broken it down for logs during testing.
